Cassandra 20402 trunk

ekaterinadimitrova2 opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

ekaterinadimitrova2 committed:
CASSANDRA-20402: Add new reason RequestFailureReason.INDEX_BUILD_IN_PROGRESS

ekaterinadimitrova2 committed:
DO NOT COMMIT  CircleCI config for testing

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -316,7 +317,13 @@ public void checkQueryability(Index.QueryPlan queryPlan)
             if (!isIndexQueryable(index))
+            {
+                // In Astra index can be queryable during index build, thus we need to check both not queryable and building
+                if (isIndexBuilding(index))
```
Hmmm, we probably need to use instead directly the full rebuild status here instead? Same as in `filterForQuery`

maedhroz replied:
You're right. I don't think we'd want to count the cases where we aren't doing a full/initial build.

ekaterinadimitrova2 commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexAvailabilityTest.java:
```diff
@@ -188,6 +190,92 @@ private void markIndexNonQueryable(IInvokableInstance node, String keyspace, Str
+
+            assertThatThrownBy(() ->
+                    executeOnAllCoordinators(cluster,
+                            "SELECT pk FROM " + ks2 + '.' + cf1 + " WHERE v1=0 AND v2=0"
+                    ))
+                    .hasMessageContaining("Operation failed - received 1 responses and 1 failures: INDEX_BUILD_IN_PROGRESS");
```
In 2i ALLOW FILTERING seems to pass here and this query we have here instead complains of lack of ALLOW FILTERING and not about index not being available or building

ekaterinadimitrova2 replied:
Which makes sense considering:
https://github.com/apache/cassandra/pull/3955/files#r1990063343

maedhroz commented on src/java/org/apache/cassandra/exceptions/RequestFailureReason.java:
```diff
@@ -53,26 +59,31 @@ public enum RequestFailureReason
         {
-            if (codeMap[reason.code] != null)
+            if (codeToReasonMap.put(reason.code, reason) != null)
                 throw new RuntimeException("Two RequestFailureReason-s that map to the same code: " + reason.code);
```
```suggestion
                throw new RuntimeException("Two RequestFailureReasons that map to the same code: " + reason.code);
```
Not sure what the dash was doing in there...

ekaterinadimitrova2 replied:
typo? or someone wanted to stress on the class name?

ekaterinadimitrova2 replied:
Do you want me to revert it?

maedhroz replied:
I guess if we're here, we might as well get rid of the dash.

maedhroz commented on src/java/org/apache/cassandra/exceptions/RequestFailureReason.java:
```diff
@@ -53,26 +59,31 @@ public enum RequestFailureReason
+        exceptionToReasonMap.put(IndexBuildInProgressException.class, INDEX_BUILD_IN_PROGRESS);
+
+        if (exceptionToReasonMap.size() != reasons.length-3)
+            throw new RuntimeException("A new RequestFailureReasons was probably added and you may need to update the exceptionToReasonMap");
```
What if we just add a nullable field to the enum `Class<? extends Throwable> exception` and do...

```
if (reason.exception != null)
    exceptionToReasonMap.put(reason.exception, reason);
```
...in the loop above? Seems like it helps us avoid the need for this size check, which could get brittle.

ekaterinadimitrova2 replied:
Hmmm, that makes a lot of sense. I'm not sure why I didn't do it. Honestly, I can't think of a good reason. Thanks

ekaterinadimitrova2 replied:
Actually, the size check was also to make people not forget to add exceptions. 
If we do just 
```
if (reason.exception != null)
    exceptionToReasonMap.put(reason.exception, reason);
```
we lose that... And then we can end up again with `UNKNOWN` from `forException`
I can probably add a unit test to catch that, but we are just moving it to CI and it can be still missed, plus you get the signal only when you get to CI. Earlier feedback while coding sounds better to me.
Let me know what you prefer

maedhroz replied:
That's fine. Maybe we can still get rid of the magic number "3" somehow though?

ekaterinadimitrova2 replied:
Added constant in this commit - https://github.com/apache/cassandra/commit/30a61c4501969267a192a2398dfbc4ef7d251c57

maedhroz commented on src/java/org/apache/cassandra/exceptions/RequestFailureReason.java:
```diff
@@ -36,14 +39,17 @@ public enum RequestFailureReason
-
+    // below code is intentionally higher, coming as a back port from a fork. It was created that high to prevent
+    // any conflicts with Apache Cassandra code.
+    INDEX_BUILD_IN_PROGRESS  (503);
```
I'm guessing this might not be the only one of these. Thinking of a general comment that takes that into account...maybe something like...

```
// The following codes have been ported from an external fork, where they were offset explicitly to avoid conflicts.
```

maedhroz commented on src/java/org/apache/cassandra/exceptions/RequestFailureReason.java:
```diff
@@ -81,25 +92,18 @@ public static RequestFailureReason fromCode(int code)
-            return COORDINATOR_BEHIND;
+        for (Map.Entry<Class<? extends Throwable>, RequestFailureReason> entry : exceptionToReasonMap.entrySet())
+            if (entry.getKey().isInstance(t))
+                return entry.getValue();
```
So `forException()` is actually now mapping `TIMEOUT`, `READ_SIZE`, and `INDEX_NOT_AVAILABLE` where it wasn't before, correct? Just wanted to make sure that was intended (and a little curious about why there weren't already mapped...)

ekaterinadimitrova2 replied:
Those were missed. So, I am adding the mapping and the brittle way of checking the length now to make sure people deliberately add or decide not to add them here. The drivers use this method with v5, where the `reasonMap` exists. It occurs to me that people just missed adding those; it's not hard to do.

maedhroz commented on src/java/org/apache/cassandra/index/IndexBuildInProgressException.java:
```diff
@@ -0,0 +1,34 @@
+package org.apache.cassandra.index;
+
+/**
+ * Thrown if a secondary index is not currently available because it is building.
```
nit: Still need to read the rest of the PR, but I'm guessing this specifically means the initial index build...will come back to this once I have a full understanding...

ekaterinadimitrova2 replied:
It means any full index build. I want to leave the differentiation with the initial index build for the other ticket where I am adding also particular` Index.Status`. This is small initial improvement.

maedhroz commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -316,7 +317,13 @@ public void checkQueryability(Index.QueryPlan queryPlan)
         {
             if (!isIndexQueryable(index))
+            {
+                // In Astra index can be queryable during index build, thus we need to check both not queryable and building
```
nit: I guess we should eventually remove the specific reference to Astra?

ekaterinadimitrova2 replied:
Actually, it seems it applies to any non-SAI index implementation. Looking into [`isQueryable()`](https://github.com/datastax/cassandra/blob/main/src/java/org/apache/cassandra/index/Index.java#L496-L499) (which is used to decide if we add to `queryableIndexes`) - it returns always true for non-SAI implementations.

maedhroz replied:
Yeah, for some historical reason `CassandraIndex` looks like...

```
return status == Status.BUILD_SUCCEEDED || status == Status.UNKNOWN || status == Status.FULL_REBUILD_STARTED;
```

...whereas SAI won't say it's queryable during a full rebuild.

ekaterinadimitrova2 replied:
Yeah, my bad, CassandraIndex overrides it too since recently, but still, only BUILD_FAILED and DROPPED means non-queryable

ekaterinadimitrova2 replied:
So something like that? (with non-destroyed by GH formatting, of course)
```
public void checkQueryability(Index.QueryPlan queryPlan)
    {
        InetAddressAndPort endpoint = FBUtilities.getBroadcastAddressAndPort();

        for (Index index : queryPlan.getIndexes())
        {
            String indexName = index.getIndexMetadata().name;
            Index.Status indexStatus = getIndexStatus(endpoint, keyspace.getName(), indexName);

            if (!isIndexQueryable(index))
            {
                // Plus isQueryable is always true for non-SAI index implementations,  thus we need to check both not queryable and building
                if (indexStatus == Index.Status.FULL_REBUILD_STARTED)
                    throw new IndexBuildInProgressException(index);

                throw new IndexNotAvailableException(index);
            }
        }
    }
```

maedhroz replied:
Yeah, that looks reasonable

maedhroz commented on test/unit/org/apache/cassandra/index/sai/cql/AllowFilteringTest.java:
```diff
@@ -18,13 +18,17 @@
 
+import org.apache.cassandra.index.IndexBuildInProgressException;
+import org.apache.cassandra.inject.Injections;
+import org.apache.cassandra.inject.InvokePointBuilder;
```
nit: move these C* imports down

maedhroz commented on test/unit/org/apache/cassandra/index/sai/cql/AllowFilteringTest.java:
```diff
@@ -391,4 +395,30 @@ private void test(String query, boolean requiresAllowFiltering) throws Throwable
+        blockIndexBuild.disable();
+        waitForIndexQueryable(idx);
+        execute("SELECT * FROM %s WHERE v=0");
+        execute("SELECT * FROM %s WHERE v=0 ALLOW FILTERING");
```
The point of having AF here even when it's not required is to illustrate that even w/ AF, we'll fail during the initial index build?

ekaterinadimitrova2 replied:
Yes, that nothing changed in behavior until the other ticket with ALLOW FILTERING used during building is solved

maedhroz commented on test/unit/org/apache/cassandra/index/sai/cql/AllowFilteringTest.java:
```diff
@@ -391,4 +395,30 @@ private void test(String query, boolean requiresAllowFiltering) throws Throwable
+        String idx = createIndexAsync(String.format("CREATE CUSTOM INDEX ON %%s(v) USING '%s'", StorageAttachedIndex.class.getName()));
+
+        assertThatThrownBy(() -> execute("SELECT * FROM %s WHERE v=0"))
+                .hasMessage("The secondary index '" + idx + "' is not yet available as it is building")
```
I wonder if we could make a little constant out of the message string/pattern in `IndexBuildInProgressException`? That way we could reuse it here (and in some other tests?) to build error messages without worrying if we'll change them later (assuming they still take just the index name).

ekaterinadimitrova2 replied:
I have it and I forgot to commit it :-) Good call

maedhroz commented on test/unit/org/apache/cassandra/exceptions/RequestFailureReasonTest.java:
```diff
@@ -54,4 +57,48 @@ public void testEnumCodesAndNames()
+        assertEquals(RequestFailureReason.NOT_CMS, RequestFailureReason.fromCode(8));
+        assertEquals(RequestFailureReason.INVALID_ROUTING, RequestFailureReason.fromCode(9));
+        assertEquals(RequestFailureReason.COORDINATOR_BEHIND, RequestFailureReason.fromCode(10));
+        assertEquals(RequestFailureReason.INDEX_BUILD_IN_PROGRESS, RequestFailureReason.fromCode(503));
```
nit: Can we just use `EXPECTED_VALUES` to test these in a loop? At least that way we only have to modify `EXPECTED_VALUES` when we add a new code...

maedhroz commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -25,6 +25,8 @@
 
 import com.google.common.collect.ImmutableSet;
+
+import org.apache.cassandra.index.IndexBuildInProgressException;
```
nit: move import down

maedhroz commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexAvailabilityTest.java:
```diff
@@ -27,6 +27,7 @@
 import java.util.concurrent.TimeUnit;
 
 import com.google.common.base.Objects;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
```
nit: move import down

maedhroz commented on test/unit/org/apache/cassandra/index/sai/cql/AllowFilteringTest.java:
```diff
@@ -391,4 +395,30 @@ private void test(String query, boolean requiresAllowFiltering) throws Throwable
 
+    private final Injections.Barrier blockIndexBuild = Injections.newBarrier("block_index_build", 2, false)
+                                                                 .add(InvokePointBuilder.newInvokePoint().onClass(StorageAttachedIndex.class)
+                    .onMethod("startInitialBuild"))
```
nit: formatting?

ekaterinadimitrova2 committed:
Address nits from review, check only for full index builds. SecondaryIndexManager#markIndexFailed should make the index non-queryable with Index.Status.BUILD_FAILED reason.

ekaterinadimitrova2 commented:
@maedhroz , I addressed the nits from review and the check only for full index builds. The latter led to - SecondaryIndexManager#markIndexFailed should make the index non-queryable with Index.Status.BUILD_FAILED reason. I also rebased the branch. 
https://github.com/apache/cassandra/pull/3955#discussion_r2014136069 - I believe this is the only last question. Please let me know what you prefer and I will address it and re-run CI.

ekaterinadimitrova2 committed:
Address review feedback

ekaterinadimitrova2 committed:
Fix unit tests for the new exception
Account for the new exception in InboundSink

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/index/SecondaryIndexManager.java:
```diff
@@ -847,8 +847,6 @@ private synchronized void markIndexFailed(Index index, boolean isInitialBuild)
             if (!index.getSupportedLoadTypeOnFailure(isInitialBuild).supportsReads() && queryableIndexes.remove(indexName))
                 logger.info("Index [{}] became not-queryable because of failed build.", indexName);
-
-            makeIndexNonQueryable(index, Index.Status.BUILD_FAILED);
```
This doesn't work as recently we overrode the `isQueryable` for `CassandraIndex`, and `BUILD_FAILED` is a non-queryable index, but `FULL_REBUILD_STARTED` is queryable. So, we would introduce a change of behavior with this call, which we do not want to do.

ekaterinadimitrova2 committed:
Make Java distributed test deterministic

ekaterinadimitrova2 committed:
DO NOT COMMIT, only for testing

ekaterinadimitrova2 commented:
Merged manually to trunk
